pkg/controller/clustersync/clustersync_controller.go

```diff
@@ -406,7 +410,28 @@ func (r *ReconcileClusterSync) Reconcile(ctx context.Context, request reconcile.
 
 	if cd.DeletionTimestamp != nil {
 		logger.Debug("cluster is being deleted")
-		return reconcile.Result{}, nil
+
+		clusterSync := &hiveintv1alpha1.ClusterSync{}
+		switch err := r.Get(context.Background(), request.NamespacedName, clusterSync); {
+		case apierrors.IsNotFound(err):
+			logger.Debug("no ClusterSync for cluster being deleted")
+			return reconcile.Result{}, nil
+		case err != nil:
+			logger.WithError(err).Log(controllerutils.LogLevel(err), "could not get ClusterSync")
+			return reconcile.Result{}, err
+		}
+
+		if clusterSync.DeletionTimestamp != nil {
+			logger.Debug("clustersync is also being deleted")
+			return reconcile.Result{}, r.cleanUpClusterSync(logger, clusterSync, cd)
+		} else {
+			logger.Debug("delete clustersync too")
+			if err := r.Delete(context.Background(), clusterSync); err != nil {
```

Comment:
But if the clustersync is only deleted after the clusterdeployment is fully deleted, then the reconcile loop for the clustersync spins forever with "clusterdeployment not found"

Reply:
Have you seen this happen in practice, without forcing deletions? It _should_ be impossible, since the CD won't be garbage collected until its referents (including this one) are garbage collected.

Reply:
Yes, it turns out that by default kubernetes will delete the CD first before the CS ("background" cascade) as opposed to what you think would be logical (the CS being deleted before the CD, "foreground" cascade).  Luckily we can just add a special finalizer to the CD as described here: https://pkg.go.dev/k8s.io/apimachinery/pkg/apis/meta/v1#OwnerReference.BlockOwnerDeletion

Reply:
@2uasimojo ok so it's basically impossible to use this built in finalizer (and therefore ensure that the clustersync is cleaned up) unless you explicitely pass `--cascade=foreground` to `oc delete`, which is not something we want to require always.  Therefore, I think there are two options:
1. keep the codepath to manually delete the CD (as it stands currently)
2. store enough information on the clustersync object so that it doesn't require the CD to still exist in order to clean itself up

Reply:
To be clear: Adding the finalizer manually to the CD, and then deleting the CD with `oc delete` will cause k8s to [_explicitly remove the finalizer for us_](https://github.com/kubernetes/kubernetes/blob/f30a87d517c538f22e1d84b4d7c497f57f7b760a/pkg/registry/core/namespace/storage/storage.go#L212) (again, because the default mode for deleting an object is `background`)

Reply:
TL;DR: `blockOwnerDeletion` means nothing unless you supply the right option when deleting the owner